src/Multicall3.sol

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.12;
+
+/// @title Multicall3
+/// @notice Aggregate results from multiple read-only function calls
+/// @dev Multicall & Multicall2 backwards-compatible
+/// @author Michael Elliot <mike@makerdao.com>
+/// @author Joshua Levine <joshua@makerdao.com>
+/// @author Nick Johnson <arachnid@notdot.net>
+/// @author Andreas Bigger <andreas@nascent.xyz>
+
+contract Multicall3 {
+    struct Call {
+        address target;
+        bytes callData;
+    }
+
+    struct Call3 {
+        address target;
+        bytes callData;
+        bool requireSuccess;
+    }
+
+    struct Result {
+        bool success;
+        bytes returnData;
+    }
+
+    function aggregate(Call[] memory calls) public returns (uint256 blockNumber, bytes[] memory returnData) {
+        blockNumber = block.number;
+        uint256 length = calls.length;
+        returnData = new bytes[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(success, "Multicall aggregate: call failed");
+            returnData[i] = ret;
+            unchecked { i++; }
+        }
+    }
+
+    function tryAggregate(bool requireSuccess, Call[] memory calls) public returns (Result[] memory returnData) {
+        uint256 length = calls.length;
+        returnData = new Result[](length);
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+
+            if (requireSuccess) {
+                require(success, "Multicall2 aggregate: call failed");
+            }
+
+            returnData[i] = Result(success, ret);
+
+            unchecked { i++; }
+        }
+    }
+
+    function tryBlockAndAggregate(bool requireSuccess, Call[] memory calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = tryAggregate(requireSuccess, calls);
+    }
+
+    function blockAndAggregate(Call[] memory calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        (blockNumber, blockHash, returnData) = tryBlockAndAggregate(true, calls);
+    }
+
+    function aggregate3(Call3[] memory calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = new Result[](calls.length);
+        uint256 length = calls.length;
+        for (uint256 i = 0; i < length;) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(!calls[i].requireSuccess || success, "Multicall3 aggregate3: call failed");
```

Comment:
i would change requireSuccess to allowFailure, then u can avoid an iszero

Reply:
also depending if ur expecting most ppl to use that u might wanna use bitwise or

Reply:
Sorry what do you mean by that second comment? Using bitwise or would avoid short circuiting right?

And allowFailure seems reasonable to me, should benchmark with optimizer to confirm savings

Reply:
You're distracting me from linux kernel hacking, but writing an impl now :)

Reply:
> Sorry what do you mean by that second comment? Do you mean using or() to avoid short circuiting?

ye

Reply:
> And allowFailure seems reasonable to me, should benchmark with optimizer to confirm savings

the optimizer can't remove the iszero() it needs to be there if ur negating something in a require

Reply:
Oh.. So it's better to check that something fails than to make sure it doesn't? @transmissions11